README.md
# README

form Yerdanksa
extension for Visual Studio Code.

## Features

<img src="images/command-palette.png" width="600" alt="The command palette.">
<i>The command palette.</i><br><br><br>
<img src="images/microphone-prompt.png" width="600" alt="The extension will ask you for permission to use your microphone.">
<i>The extension will ask you for permission to use your microphone.</i><br><br><br>
<img src="images/address-bar-allowed.png" width="600" alt="Address bar in Chrome when you've allowed access to your microphone.">
<i>Address bar in Chrome when you've allowed access to your microphone.</i><br><br><br>

## استفاده

1. command palette در `cmd/ctrl+shift+p` اجرای
2. command در قسمت `Yerdanksa: Dictate` اجرای
3. در آدرس مرورگر وارد کنید `localhost:9000` اگر به صورت اتوماتیک مرورگر اجرا نشد به صورت دستی مرورگر کروم را باز کنید و پورت
4. در صورتی که مرورگر اجازه دسترسی به میکروفون را خواست این اجازه را صادر کنید
5. شروع به صحبت کردن کنید

> توجه: برنامه افزودنی فقط روی فایل هایی که قبلاً روی دیسک ذخیره شده اند کار می کند.

## <a name="commands">Commands</a>

> کلمات کلیدی

| line | add new Line |
| parentheses | add "( )" |
| equal | add "=" |
| brace | add "{ }" |
| space | add " " |
| colon | add ":" |
| double quotation | add "" "" |
| single quotation | add "' '" |
| pline | Move cursor prevBlankLine. |
| nline | Move cursor nextBlankLine. |
| down | Move cursor down. |
| enter | Accept current IntelliSense suggestion. |
| left | Move cursor left. |
| next | Select next IntelliSense suggestion. |
| previous | Select previous IntelliSense suggestion. |
| right | Move cursor right. |
| up | Move cursor up. |

## الزامات

1. VS Code
2. microphone
3. Internet connection.
4. A browser that supports the Web Speech API
5. Available ports `9000` and `9001`.


CHANGELOG.md
# Change Log

All notable changes to the "Yerdanksa Rejshth" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release


src/constant/command-mappings.ts
import { ICommand } from '../interface/command';

/**
 * Map certain spoken phrases to VS Code commands.
 * Key is the spoken phrase, value the command.
 */
export const COMMAND_MAPPINGS: {[key:string]: ICommand} = {
	down: {
		command: 'cursorMove',
		params: [{
			to: 'down',
		}],
	},
	enter: {
		command: 'acceptSelectedSuggestion',
		if: true, // Todo: Check if suggestion dialog is open. If not, add new line instead.
	},
	left: {
		command: 'cursorMove',
		params: [{
			to: 'left',
		}],
	},
	next: {
		command: 'selectNextSuggestion',
	},
	
	previous: {
		command: 'selectPrevSuggestion',
	},
	right: {
		command: 'cursorMove',
		params: [{
			to: 'right',
		}],
	},
	up: {
		command: 'cursorMove',
		params: [{
			to: 'up',
		}],
	},
	space: {
		command: 'cursorMove',
		params: [{
			to: 'space',
		}],
	},
	nline: {
		command: 'cursorMove',
		params: [{
			to: 'nextBlankLine',
		}],
	},
	pline: {
		command: 'cursorMove',
		params: [{
			to: 'prevBlankLine',
		}],
	},

};

/*
print("Hello World")
-----------------------------------
movie = "Terminator"
number1 = 5
float_number = 12.5
str = "ITPro.ir python course"
-------------------
list = [item1, item2, item3, item4]
list = ["ITPro.ir", 12, 3.5]
---------------------------------
تاپل ها همانند لیست مجموعه ای از آیتم ها را تعریف می کنند، اما با لیست دو تفاوت اصلی دارند:

    آیتم های Tuple داخل () قرار میگرند.
    آیتم های Tuple قابل تغییر نیستند، یعنی Tuple ها در حقیقت Read-Only هستند.
tuple = ("Item1","Item2")
------------------------------
دیکشنری ها نوعی از داده ها هستند که بوسیله ساختار کلید و مقدار مشخص می شوند. یعنی شما برای هر آیتم دیکشنری یک کلید مشخص می کنید و یک مقدار و بعد می توانید بوسیله کلید به مقدار مورد نظر داخل دیکشنری دسترسی داشته باشید. در مثال زیر، یک دیکشنری خیلی ساده تعریف کرده ایم: 
dict = {"name":"Hossein", "age":30,"website":"ITPro.i"}
------------------------------------
def printMessage(firstName,lastName):
    message = "Welcome "+firstName+" "+lastName
  print(message)
--------------------------
public voic int PrintName(string firstName, string lastName)
{
    string message = "Welcome "+firstName+" "+lastName;
    Console.WriteLine(message);
}
-----------------------
# Source code developed by
---------------------
if expression:
    statements
else:
    statements

--------------------------
if expression:
    statements
elif expression:
    statements
elif expression:
    statements
else:
    statements

----------------------------
می توان در زبان پایتون شرط ها را به صورت تو در تو نوشت، یعنی داخل یک شرط if یک شرط if دیگر به صورت زیر تعریف کرد:

if expression:
    if expression:
        statements
    elif expression:
        statements
else
    statements
---------------------------------------
while expression:
    statement(s)


n = 1
while n <= 10:
    print(n)
    n += 1

----------------------------------------------

n = 1
while n <= 10:
    print(n)
    n += 1
else:
    print("End of loop!")
-----------------------------------------------
list = ["item1","item2","item3","item4"]

for item in list:
    print(item)

print("All done.")

-------------------------------------
class Person:
    class body


class Person:
    def __init__(self, firstname, lastname, age):
        self.firstname = firstname
        self.lastname = lastname
        self.age = age


---------------------------

def function_name(parameters):
    statement(s)

def sum(num1,num2):
    print(num1+num2)

def sum(num1,num2):
   return num1 + num2

--------------------------------



*/

src/extension.ts

import * as vscode from 'vscode';
import { COMMAND_MAPPINGS } from './constant/command-mappings';
const path = require('path');
const WebSocket = require('ws');
const express = require('express');
const open = require('open');
const app = express();
const port = 9000;
const websocketPort = 9001;

app.use('/', express.static(path.join(__dirname, 'client')));
app.listen(port, () => {
	vscode.window.showInformationMessage(`[Yerdanksa Rejshth] Server running at localhost:${port}`);
	open(`http://localhost:${port}`);
});


export function activate(context: vscode.ExtensionContext) {
	console.log('[Yerdanksa Rejshth] Congratulations, your extension "Yerdanksa Rejshth" is now active!');
	let disposable = vscode.commands.registerCommand('stt.dictate', () => {
		const wss = new WebSocket.Server({port: websocketPort});
		wss.on('connection', (socket: any) => {
			console.log('[Yerdanksa Rejshth] New WebSocket connection');
			vscode.window.showInformationMessage('[Yerdanksa Rejshth] New WebSocket connection');

			socket.on('message', (phrase: string) => {
				console.log(`[Yerdanksa Rejshth] New WebSocket message: ${phrase}`);
				vscode.window.showInformationMessage(`Speech : ${phrase}`);

				switch (phrase) {
					case "colon":
						 phrase=":";
						 break;

					case "equal":
						 phrase="=";
						 break;

					case "space":
						 phrase=" ";
						 break;

					case "parentheses": 
					phrase="( )";
					break;

					case "brace": 
					phrase="{ }";
					break;

					case "dot": 
					phrase=".";
						break;
				
						case "double quotation": 
					phrase="\" \"";
						break;
					
						case "single quotation": 
						phrase="' '";
							break;
						
							case "line": 
							phrase="\n";
								break;
					default:
						break;
				}

			
				// if (phrase==="class") {
				// 	phrase="class Person:\n"

				// }
				// if(phrase==="colon") phrase=":"
				// if(phrase==="equal") phrase="="
				
				
				const path = vscode.window.activeTextEditor?.document.fileName;
				
				if (path) {
					if (
						Object.keys(COMMAND_MAPPINGS).includes(phrase) && 
						(
							(Object.keys(COMMAND_MAPPINGS[phrase]).includes('if') && COMMAND_MAPPINGS[phrase].if) || 
							!Object.keys(COMMAND_MAPPINGS[phrase]).includes('if')
						)
					) {
						vscode.commands.executeCommand(COMMAND_MAPPINGS[phrase].command, ...(COMMAND_MAPPINGS[phrase].params || []));
					} else {
						const edit = new vscode.WorkspaceEdit();
						const uri = vscode.Uri.file(path);
						const cursorPos = vscode.window.activeTextEditor?.selection.active;
						const position = new vscode.Position(cursorPos?.line||0, cursorPos?.character||0);
						edit.insert(uri, position, phrase);
						vscode.workspace.applyEdit(edit).then(res => vscode.commands.executeCommand('editor.action.triggerSuggest'));
					}
				}
			});
		});
	});

	context.subscriptions.push(disposable);
}

export function deactivate() {}
